src/v/kafka/server/group_manager.cc

```diff
@@ -403,29 +461,31 @@ ss::future<> group_manager::do_detach_partition(model::ntp ntp) {
         co_return;
     }
     auto p = it->second;
-    auto units = co_await p->catchup_lock.hold_write_lock();
-
-    // Becasue shutdown group is async operation we should run it after
-    // rehash for groups map
-    std::vector<group_ptr> groups_for_shutdown;
-    for (auto g_it = _groups.begin(); g_it != _groups.end();) {
-        if (g_it->second->partition()->ntp() == p->partition->ntp()) {
-            groups_for_shutdown.push_back(g_it->second);
-            _groups.erase(g_it++);
-            continue;
+
+    co_await with_wlock("do_detach_partition", p, [this, p]() mutable {
+        // Because shutdown group is async operation we should run it after
+        // rehash for groups map
+        std::vector<group_ptr> groups_for_shutdown;
+        for (auto g_it = _groups.begin(); g_it != _groups.end();) {
+            if (g_it->second->partition()->ntp() == p->partition->ntp()) {
+                groups_for_shutdown.push_back(g_it->second);
+                _groups.erase(g_it++);
+                continue;
+            }
+            ++g_it;
         }
-        ++g_it;
-    }
-    // if p has background work that won't complete without an abort being
-    // requested, then do that now because once the partition is removed from
-    // the _partitions container it won't be available in group_manager::stop.
-    if (!p->as.abort_requested()) {
-        p->as.request_abort();
-    }
-    _partitions.erase(ntp);
-    _partitions.rehash(0);
+        // if p has background work that won't complete without an abort being
+        // requested, then do that now because once the partition is removed
+        // from the _partitions container it won't be available in
+        // group_manager::stop.
+        if (!p->as.abort_requested()) {
+            p->as.request_abort();
+        }
+        _partitions.erase(p->partition->ntp());
+        _partitions.rehash(0);
 
-    co_await shutdown_groups(std::move(groups_for_shutdown));
```

Comment:
I have a question.. why should shutdown_groups() be within catchup_writelock scope? Does the following break something? 

```

std::vector<group_ptr> groups_for_shutdown;
{
   auto units = co_await p->catchup_lock.hold_write_lock();

    // Becasue shutdown group is async operation we should run it after
    // rehash for groups map
 
    for (auto g_it = _groups.begin(); g_it != _groups.end();) {
        if (g_it->second->partition()->ntp() == p->partition->ntp()) {
            groups_for_shutdown.push_back(g_it->second);
            _groups.erase(g_it++);
            continue;
        }
        ++g_it;
    }
    // if p has background work that won't complete without an abort being
    // requested, then do that now because once the partition is removed from
    // the _partitions container it won't be available in group_manager::stop.
    if (!p->as.abort_requested()) {
        p->as.request_abort();
    }
    _partitions.erase(ntp);
    _partitions.rehash(0);
}
co_await shutdown_groups(std::move(groups_for_shutdown));
```

Reply:
Most likely nothing but it takes effort to reason so I prefer to leave it as is. The PR doesn't introduce this lock there so I think it's safer to left it intact.

Reply:
I think this is related to the deadlock.. AFAICT, moving shutdown_groups() out of write scope removes the nested lock/gate dependency and fixes the deadlock, worth exploring.